File: C/Data Structures/Linked Lists/problem21.c
/* Problem 2.1
 * Write code to remove duplicates from an unsorted linked list. Temporary buffer is not allowed.
 *
 * Solution
 * - Time complexity: O(n^2).
 * - Space complexity: O(1)
 * If we use a temporary buffer we will be able to reach time and space complexities both equal to O(n).
 * Look at the Java implementation for this solution.
 */

#include "stdio.h"
#include "lists.h"

void delete_duplicates(list l);

int main() {
    list l = NULL;
    l = tail_insert(l, 3);
    l = tail_insert(l, 2);
    l = tail_insert(l, 3);
    l = tail_insert(l, 1);
    l = tail_insert(l, 1);
    l = tail_insert(l, 5);
    printf("Starting list: ");
    print_list(l);
    delete_duplicates(l);
    printf("Ending list: ");
    print_list(l);

    return 0;
}

void delete_duplicates(list l) {
    list temp = l, cur = NULL;
    while(temp != NULL) {
        cur = temp;
        while(cur -> next != NULL) {
            if(temp -> data == (cur -> next) -> data)
                cur -> next = (cur -> next) -> next;
            else
                cur = cur -> next;
        }
        temp = temp -> next;
    }
}

File: Python/Data Structures/Linked Lists/problem21.py
# Problem 2.1
# Write code to remove duplicates from an unsorted linked list. Temporary buffer is not allowed.
#
# Solution
# - Time complexity: O(n^2).
# - Space complexity: O(1)
# If we use a temporary buffer we will be able to reach time and space complexities both equal to O(n).
# Look at the Java implementation for this solution.

import LinkedList as ll


def delete_duplicates(linked_list):
    temp = linked_list.head
    while temp is not None:
        cur = temp
        while cur.next is not None:
            if temp.data == cur.next.data:
                cur.next = cur.next.next
            else:
                cur = cur.next
        temp = temp.next


l = ll.LinkedList()
l.head_insert(4)
l.head_insert(4)
l.head_insert(2)
l.head_insert(0)
l.head_insert(2)
print("Starting list:", end=" ")
l.print_list()
delete_duplicates(l)
print("Ending list:", end=" ")
l.print_list()
